File: modules/vescape-core/android/src/main/java/expo/modules/vescapecore/location/LegalPolicyResolver.kt
package expo.modules.vescapecore.location

import android.content.Context
import android.location.Geocoder
import java.util.Locale
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext

/**
 * Native OS reverse geocoder constrained by the bundled Legal Policy catalog.
 *
 * @parity /modules/vescape-core/ios/location/LegalPolicyResolver.swift
 */
internal class LegalPolicyResolver(private val context: Context) {
    private val catalog = LegalPolicyCatalog(context)

    @Suppress("DEPRECATION")
    suspend fun resolve(latitude: Double, longitude: Double): String? = withContext(Dispatchers.IO) {
        val address = runCatching {
            Geocoder(context, Locale.getDefault()).getFromLocation(latitude, longitude, 1)?.firstOrNull()
        }.getOrNull()
        normalizeCountryCode(address?.countryCode, catalog.countryCodes)
    }
}

internal fun normalizeCountryCode(raw: String?, supported: Set<String>): String? {
    val code = raw?.trim()?.uppercase(Locale.ROOT) ?: return null
    return code.takeIf(supported::contains)
}

File: modules/vescape-core/ios/location/LegalPolicyResolver.swift
import CoreLocation
import Foundation

/// Native OS reverse geocoder constrained by the bundled Legal Policy catalog.
///
/// @parity /modules/vescape-core/android/src/main/java/expo/modules/vescapecore/location/LegalPolicyResolver.kt
internal final class LegalPolicyResolver {
  private let geocoder = CLGeocoder()
  private lazy var supportedCountryCodes = Self.bundledCountryCodes()

  func resolve(latitude: Double, longitude: Double) async -> String? {
    let placemarks = try? await geocoder.reverseGeocodeLocation(
      CLLocation(latitude: latitude, longitude: longitude)
    )
    return Self.normalizeCountryCode(placemarks?.first?.isoCountryCode, supported: supportedCountryCodes)
  }

  static func countryCodes(json: String) -> Set<String> {
    guard
      let data = json.data(using: .utf8),
      let rows = try? JSONSerialization.jsonObject(with: data) as? [[String: Any]]
    else { return [] }
    return Set(rows.compactMap { ($0["code"] as? String)?.trimmingCharacters(in: .whitespaces).uppercased() })
  }

  static func normalizeCountryCode(_ raw: String?, supported: Set<String>) -> String? {
    guard let code = raw?.trimmingCharacters(in: .whitespaces).uppercased() else { return nil }
    return supported.contains(code) ? code : nil
  }

  private static func bundledCountryCodes() -> Set<String> {
    let moduleBundle = Bundle(for: LegalPolicyResolver.self)
    let url =
      moduleBundle.url(forResource: "legal-policies", withExtension: "json")
      ?? moduleBundle.url(forResource: "VescapeCoreAssets", withExtension: "bundle").flatMap {
        Bundle(url: $0)?.url(forResource: "legal-policies", withExtension: "json")
      }
    guard let url, let json = try? String(contentsOf: url, encoding: .utf8) else { return [] }
    return countryCodes(json: json)
  }
}
